File: solutions/python/04/segtree/main.py
class SegTree:
    def __init__(self, data):
        n = 1
        while n < len(data):
            n <<= 1
        self.n = n
        self.seg = [0] * (2 * n)
        for i, v in enumerate(data):
            self.seg[n + i] = v
        for i in range(n - 1, 0, -1):
            self.seg[i] = self.seg[2 * i] + self.seg[2 * i + 1]

    def update(self, i, x):
        i += self.n
        self.seg[i] = x
        while i > 1:
            i >>= 1
            self.seg[i] = self.seg[2 * i] + self.seg[2 * i + 1]

    def query(self, l, r):
        l += self.n
        r += self.n
        res = 0
        while l < r:
            if l & 1:
                res += self.seg[l]
                l += 1
            if r & 1:
                r -= 1
                res += self.seg[r]
            l >>= 1
            r >>= 1
        return res


def main() -> None:
    st = SegTree([1, 2, 3, 4, 5])
    print(st.query(1, 4))
    st.update(2, 10)
    print(st.query(1, 4))


if __name__ == "__main__":
    main()

File: solutions/go/04/segtree/main.go
package main

import "fmt"

type SegTree struct {
	n   int
	seg []int
}

func NewSegTree(data []int) *SegTree {
	n := 1
	for n < len(data) {
		n <<= 1
	}
	seg := make([]int, 2*n)
	for i, v := range data {
		seg[n+i] = v
	}
	for i := n - 1; i > 0; i-- {
		seg[i] = seg[2*i] + seg[2*i+1]
	}
	return &SegTree{n, seg}
}

func (s *SegTree) Update(i, x int) {
	i += s.n
	s.seg[i] = x
	for i > 1 {
		i >>= 1
		s.seg[i] = s.seg[2*i] + s.seg[2*i+1]
	}
}

func (s *SegTree) Query(l, r int) int {
	l += s.n
	r += s.n
	res := 0
	for l < r {
		if l&1 == 1 {
			res += s.seg[l]
			l++
		}
		if r&1 == 1 {
			r--
			res += s.seg[r]
		}
		l >>= 1
		r >>= 1
	}
	return res
}

func main() {
	st := NewSegTree([]int{1, 2, 3, 4, 5})
	fmt.Println(st.Query(1, 4))
	st.Update(2, 10)
	fmt.Println(st.Query(1, 4))
}
